### PR TITLE
self_delegation_record_metrics

### PR DESCRIPTION
**Pull request by pjdurden**
Self delegation field added to the validator metrics. 
Added changes to compute self_delegated_amount for a validator.
created tests for the compute_current_metrics and record_metrics functions

### CODE REVIEW

File: src/contract.rs [hunk 1]
```diff
@@ -206,6 +216,7 @@ fn add_validator(
     deps: DepsMut,
     info: MessageInfo,
     validator_addr: Addr,
+    account_addr: Addr, // account address of the validator
```
Comment (galacticminter) [inline src/contract.rs:219]: Make this a string.

Verify using deps.api.addr_validate() on this string.

Comment (galacticminter) [reply src/contract.rs:219]: Let's also call this a wallet_address

File: src/contract.rs [hunk 2]
```diff
@@ -270,8 +289,8 @@ pub fn record_validator_metrics(
     if info.sender != manager {
         return Err(ContractError::Unauthorized {});
     }
-
-    let validators_to_record = get_validators_to_record(deps.storage, timestamp)?;
+    let validator_data = get_validators_to_record(deps.storage, timestamp)?;
+    let validators_to_record = validator_data;
```
Comment (galacticminter) [inline src/contract.rs:293]: Why was this change required?

File: src/contract.rs [hunk 3]
```diff
@@ -353,10 +375,41 @@ fn compute_current_metrics(
...
+        };
+
+        // this stores current_slashing pointer value
+        // if there are no prev metric then it's default value is 1.0
+        let mut current_slashing_pointer = Decimal::one();
+
+        // formula=(current_delegated_amount/prev_delegated_amount)*prev_slashing_pointer
+        if !current_metrics.is_empty() {
+            let most_recent_metric = current_metrics.last().unwrap();
+            current_slashing_pointer = decimal_division_in_256(
```
Comment (galacticminter) [inline src/contract.rs:396]: call this a delegation_change_ratio

File: src/contract.rs [hunk 4]
```diff
@@ -353,10 +375,41 @@ fn compute_current_metrics(
...
+            current_slashing_pointer = decimal_multiplication_in_256(
+                current_slashing_pointer,
+                most_recent_metric.slashing_pointer,
+            );
+        }
+
         current_metrics.push(ValidatorMetrics {
-            addr: validator_addr.clone(),
+            operator_addr: validator_addr.operator_address.clone(),
+            account_addr: validator_addr.account_address.clone(),
```
Comment (galacticminter) [inline src/contract.rs:408]: probably don't need account & operator_addr in ValidatorMetrics. Do we?

**Review by galacticminter — APPROVED** [review]

File: src/contract.rs [hunk 5]
```diff
@@ -353,15 +367,28 @@ fn compute_current_metrics(
...
         let current_delegated_amount = delegation.amount.amount.clone();
 
+        let self_delegation_opt = deps.querier.query_delegation(
+            validators_account_addr[ind_acc_addr].clone(), //This is the Account Address
+            validator_addr.clone(),                        //This is the Operator Address
+        );
+
+        // This is the self_delegation amount (delegation by validator)
+        let self_delegation = match self_delegation_opt {
+            Ok(delegationresult) => delegationresult.unwrap().amount.amount,
```
[the remark this answers is not in the document]
Comment (pjdurden) [reply src/contract.rs]: Converted delegationresult variable's casing to snake case.

File: src/contract.rs [hunk 6]
```diff
@@ -353,15 +367,28 @@ fn compute_current_metrics(
         // This is the new Delegated amount after slashing Ex: (10 => 9.8 etc.,)
         let current_delegated_amount = delegation.amount.amount.clone();
 
+        let self_delegation_opt = deps.querier.query_delegation(
+            validators_account_addr[ind_acc_addr].clone(), //This is the Account Address
+            validator_addr.clone(),                        //This is the Operator Address
+        );
+
+        // This is the self_delegation amount (delegation by validator)
+        let self_delegation = match self_delegation_opt {
```
[the remark this answers is not in the document]
Comment (pjdurden) [reply src/contract.rs]: renamed self_delegation to self_delegation_amount. Also used match expression to set self_delegation_amount to 0 if no self_delegation done by validator.

File: src/contract.rs [hunk 7]
```diff
@@ -353,15 +367,28 @@ fn compute_current_metrics(
...
             addr: validator_addr.clone(),
             rewards: decimal_summation_in_256(current_rewards_diff, previous_rewards),
             delegated_amount: current_delegated_amount,
+            self_delegated_amount: self_delegation,
             commission: validator.commission,
             max_commission: validator.max_commission,
             rewards_in_coins: delegation.accumulated_rewards.clone(),
             timestamp,
         });
+        ind_acc_addr = ind_acc_addr + 1;
```
[the remark this answers is not in the document]
Comment (pjdurden) [reply src/contract.rs]: created new object in state named ValidatorsAccounts with operator and account address for a validator. Also fixed errors incurred after creating new object.

File: src/contract.rs [hunk 8]
```diff
@@ -435,12 +462,13 @@ fn get_total_rewards_in_vault_denom(
 fn get_validators_to_record(
     storage: &mut dyn Storage,
     timestamp: u64,
-) -> Result<Vec<Addr>, ContractError> {
+) -> Result<(Vec<Addr>, Vec<Addr>), ContractError> {
```
[the remark this answers is not in the document]
Comment (pjdurden) [reply src/contract.rs]: Changed this to a vector of Validator Accounts instead of tuple of two vectors of operator address and account address.

File: src/contract.rs (hunk 1 above)
Comment (pjdurden) [reply src/contract.rs]: Renamed account address to wallet_address and converted it from Addr to String and used deps.api.validate to validate the string wherever it is required.

File: src/contract.rs (hunk 2 above)
Comment (pjdurden) [reply src/contract.rs]: Removed the temporary variable and instead saved result of validators_to_record() to variable validators_to_record. Earlier was copying the result to a temporary variable and then copying it to the variable validators_to_record. Also fixed some lint errors.

File: src/contract.rs (hunk 3 above)
Comment (pjdurden) [reply src/contract.rs:399]: Renamed most_recent_metric to delegation_change_ratio. Also created a vector vector_delegation_change_ratio to save the previous metric of the current validator that we acquire by a call to the function query_all_validator_metrics.

File: src/contract.rs (hunk 4 above)
Comment (pjdurden) [reply src/contract.rs]: removed account_addr from Validator_Metrics. Fixed some lint errors and test suite errors.
